packages/blocks/src/root-block/clipboard/middlewares/paste.ts

```diff
@@ -148,62 +174,139 @@ class PasteTr {
-      this._getDeltas();
-
-    firstTextSnapshot.delta = [...fromDelta, ...firstDelta, ...toDelta];
+    const { firstDelta } = this._getDeltas();
+    this.fromPointState.text.applyDelta([
+      { retain: this.fromPointState.point.index },
+      this.fromPointState.point.length
+        ? { delete: this.fromPointState.point.length }
+        : {},
+      ...firstDelta,
+    ]);
+    this.snapshot.content.splice(0, 1);
+    this._updateSnapshot();
   };
 
   private _mergeMultiple = () => {
-    this.firstSnapshot.flavour = this.fromPointState.model.flavour;
+    this.firstSnapshot!.flavour = this.fromPointState.model.flavour;
     if (
-      this.firstSnapshot.props.type &&
+      this.firstSnapshot!.props.type &&
       (this.fromPointState.text.length > 0 || this.firstSnapshotIsPlainText)
     ) {
-      this.firstSnapshot.props.type = (
+      this.firstSnapshot!.props.type = (
         this.fromPointState.model as ParagraphBlockModel
       ).type;
     }
-    if (this.lastSnapshot.props.type && this.to) {
-      this.lastSnapshot.flavour = this.endPointState.model.flavour;
-      this.lastSnapshot.props.type = (
+    if (this.lastSnapshot!.props.type && this.to) {
+      this.lastSnapshot!.flavour = this.endPointState.model.flavour;
+      this.lastSnapshot!.props.type = (
         this.endPointState.model as ParagraphBlockModel
       ).type;
     }
 
-    const {
-      firstTextSnapshot,
-      lastTextSnapshot,
-      fromDelta,
-      toDelta,
-      firstDelta,
-      lastDelta,
-    } = this._getDeltas();
+    const { lastTextSnapshot, toDelta, firstDelta, lastDelta } =
+      this._getDeltas();
+
+    this.fromPointState.text.applyDelta([
+      { retain: this.fromPointState.point.index },
+      this.fromPointState.text.length - this.fromPointState.point.index
+        ? {
+            delete:
+              this.fromPointState.text.length - this.fromPointState.point.index,
+          }
+        : {},
+      ...firstDelta,
+    ]);
+
+    const removedFirstSnapshot = this.snapshot.content.shift();
+    removedFirstSnapshot?.children.map(block => {
+      this.snapshot.content.unshift(block);
+    });
+    this.pasteStartModelChildrenCount =
+      removedFirstSnapshot?.children.length ?? 0;
+
+    this._updateSnapshot();
 
-    firstTextSnapshot.delta = [...fromDelta, ...firstDelta];
     lastTextSnapshot.delta = [...lastDelta, ...toDelta];
   };
 
+  private _updateSnapshot = () => {
+    if (this.snapshot.content.length === 0) {
+      this.firstSnapshot = this.lastSnapshot = undefined;
+      return;
+    }
+    this.firstSnapshot = this.snapshot.content[0];
+    this.lastSnapshot = findLast(
+      this.snapshot.content[this.snapshot.content.length - 1]
+    );
+  };
+
+  private async _transformLinkDelta(
+    delta: DeltaOperation[],
+    linkToDocId: Map<string, string | null>,
+    quickSearchService: QuickSearchService
+  ) {
+    const needToConvert = new Map<DeltaOperation, string>();
+    for (const op of delta) {
```

Comment:
Let's use `forEach` here.

Reply:
According to [`no-misused-promises`](https://typescript-eslint.io/rules/no-misused-promises/):
```ts
// incorrect
[1, 2, 3].forEach(async value => {
  await fetch(`/${value}`);
});

// correct
for (const value of [1, 2, 3]) {
  await doSomething(value);
}
```

Reply:
Didn't notice the await here, sorry